velociraptor/catalogue/soap_catalogue.py

```diff
@@ -0,0 +1,266 @@
+import numpy as np
+import h5py
+import unyt
+
+from typing import List, Set
+from types import SimpleNamespace
+
+from velociraptor.catalogue.catalogue import Catalogue
+from velociraptor.catalogue.translator import VR_to_SOAP
+
+from functools import reduce
+
+from astropy.cosmology import wCDM, FlatLambdaCDM
+import unyt
+
+
+class CatalogueElement(object):
+    file_name: str
+    name: str
+
+    def __init__(self, file_name, name):
+        self.file_name = file_name
+        self.name = name
+
+
+class CatalogueDataset(CatalogueElement):
+
+    conversion_factor: unyt.unyt_quantity
+    value: unyt.unyt_array
+
+    def __init__(self, file_name, name, handle):
+        super().__init__(file_name, name)
+
+        self.conversion_factor = None
+        self._value = None
+        self._register_metadata(handle)
+
+    def _register_metadata(self, handle):
+        metadata = handle[self.name].attrs
+        factor = (
+            metadata["Conversion factor to CGS (including cosmological corrections)"][0]
+            * unyt.A ** metadata["U_I exponent"][0]
+            * unyt.cm ** metadata["U_L exponent"][0]
+            * unyt.g ** metadata["U_M exponent"][0]
+            * unyt.K ** metadata["U_T exponent"][0]
+            * unyt.s ** metadata["U_t exponent"][0]
+        )
+        self.conversion_factor = unyt.unyt_quantity(factor)
+        self.conversion_factor.convert_to_base("galactic")
+
+    def get_value(self, group):
+        if self._value is None:
+            with h5py.File(self.file_name, "r") as handle:
+                self._value = handle[self.name][:] * self.conversion_factor
+            self._value.name = self.name.replace("/", " ").replace("_", "").strip()
+        return self._value
+
+    def set_value(self, value, group):
+        self._value = value
+
+    def del_value(self, group):
+        del self._value
+        self._value = None
+
+
+class CatalogueDerivedDataset(CatalogueElement):
+
+    terms: List[CatalogueDataset]
+
+    def __init__(self, file_name, name, terms):
+        super().__init__(file_name, name)
+        self.terms = list(terms)
+        self._value = None
+
+    def get_value(self, group):
+        if self._value is None:
+            values = [term.get_value(group) for term in self.terms]
+            self._value = self.compute_value(*values)
+        return self._value
+
+    def set_value(self, value, group):
+        self._value = value
+
+    def del_value(self, group):
+        del self._value
+        self._value = None
+
+    def compute_value(self, *values):
+        raise NotImplementedError("This calculation has not been implemented!")
+
+
+class VelocityDispersion(CatalogueDerivedDataset):
+    def compute_value(self, velocity_dispersion_matrix):
+        return np.sqrt(
+            velocity_dispersion_matrix[:, 0]
+            + velocity_dispersion_matrix[:, 1]
+            + velocity_dispersion_matrix[:, 2]
+        )
```

Comment:
This is incorrect. `velocity_dispersion_matrix` is a 6 element array and we only want to sum the first 3 elements. But you are right that this can be done in a single line.